File: WPF/MVVM/MvvmExtensions.vb
Imports System.Runtime.CompilerServices
Imports Ks.Common.Controls

Namespace Common.MVVM

    Public Module MvvmExtensions

        <Extension()>
        Public Function IsNavigation(ByVal Self As ViewModel) As Boolean
            Return TypeOf Self Is NavigationViewModel
        End Function

        <Extension()>
        Friend Function GetNavigationView(ByVal Self As ViewModel) As INavigationView
            Return DirectCast(Self.View, INavigationView)
        End Function

        <Extension()>
        Friend Sub SetView(ByVal Navigation As ViewModel, ByVal ViewModel As ViewModel)
            Navigation.GetNavigationView().SetContent(DirectCast(ViewModel?.View, Page))
        End Sub

        <Extension()>
        Friend Sub SetContent(ByVal NavigationView As INavigationView, ByVal View As Page)
            Dim Prev = NavigationView.Content
            If Prev Is View Then
                Exit Sub
            End If

            If Prev IsNot Nothing Then
                Prev.ParentView = Nothing
            End If

            If View IsNot Nothing Then
                If View.ParentView IsNot Nothing Then
                    View.ParentView.Content = Nothing
                End If
                View.ParentView = NavigationView
            End If

            NavigationView.Content = View
        End Sub

    End Module

End Namespace

File: WPF/MVVM/MvvmExtensions.cs
using System.Windows.Controls;

using Ks.Common.Controls;

using Page = Ks.Common.Controls.Page;

namespace Ks.Common.MVVM
{
    public static class MvvmExtensions
    {
        public static bool IsNavigation(this ViewModel Self)
        {
            return Self is NavigationViewModel;
        }

        internal static INavigationView GetNavigationView(this ViewModel Self)
        {
            return (INavigationView) Self.View;
        }

        internal static void SetView(this ViewModel Navigation, ViewModel ViewModel)
        {
            Navigation.GetNavigationView().SetContent((Page) ViewModel?.View);
        }

        internal static void SetContent(this INavigationView NavigationView, Page View)
        {
            var Prev = NavigationView.Content;
            if (Prev == View)
            {
                return;
            }

            if (Prev != null)
            {
                Prev.ParentView = null;
            }

            if (View != null)
            {
                if (View.ParentView != null)
                {
                    View.ParentView.Content = null;
                }

                View.ParentView = NavigationView;
            }

            NavigationView.Content = View;
        }
    }
}
